bash: top level command completions for flux

garrettbslone opened the pull request:
This is a work in progress pr. It provides command completion for the top-level commands for flux-core in bash. This feature was requested in issue #1647. It still needs to be integrated into the makefile to be properly installed. For now, the completions can be sourced directly from flux-core/etc/completions/flux-core.

garrettbslone commented on etc/completions/flux-core:
```diff
@@ -0,0 +1,45 @@
+ #
+ # Copyright 2014 Lawrence Livermore National Security, LLC
```
I'm running bash version 4.4.19(1)-release (x86_64-pc-linux-gnu). It may also be that the extended pattern matching is not on. Try running `shopt -s extglob` to enable the extended glob that's failing.

grondo commented on etc/completions/flux-core:
```diff
@@ -0,0 +1,45 @@
+    longopts="--parent --verbose --version"
+    comopts="broker content cron dmesg env event exec get \
+             set lsattr hwloc jobs keygen kvs logger mini \
+             module ping proxy start user version help"
```
We should try to generate this list of subcommands automatically. There are a couple solutions that come to mind, feel free to try a few things out

 * run a `find` on `FLUX_EXEC_PATH` to get a list of all executables that start with `flux-`, e.g.
```
 (IFS=:; for f in $FLUX_EXEC_PATH; do find $f -maxdepth 1 -type f -executable -name flux-\* ; done)
/g/g0/grondo/git/f.git/src/broker/flux-broker
/g/g0/grondo/git/f.git/src/cmd/flux-cron
/g/g0/grondo/git/f.git/src/cmd/flux-jobspec.py
/g/g0/grondo/git/f.git/src/cmd/flux-mini.py
/g/g0/grondo/git/f.git/src/cmd/flux-keygen
/g/g0/grondo/git/f.git/src/cmd/flux-event
/g/g0/grondo/git/f.git/src/cmd/flux-logger
/g/g0/grondo/git/f.git/src/cmd/flux-aggregate
/g/g0/grondo/git/f.git/src/cmd/flux-comms
/g/g0/grondo/git/f.git/src/cmd/flux-module
/g/g0/grondo/git/f.git/src/cmd/flux-ping
/g/g0/grondo/git/f.git/src/cmd/flux-queue
/g/g0/grondo/git/f.git/src/cmd/flux-exec
/g/g0/grondo/git/f.git/src/cmd/flux-start
/g/g0/grondo/git/f.git/src/cmd/flux-kvs
/g/g0/grondo/git/f.git/src/cmd/flux-job
/g/g0/grondo/git/f.git/src/cmd/flux-jobs.py
```
(you'll have to strip the leading `*flux-` and any trailing `.py`).
This has the benefit of including site-installed subcommands, or subcommands that were added by flux-framework projects.

This won't list builtin commands though, so those will have to be added at build time (there's a script `src/cmd/builtin-cmds-list.sh` that could be modified to get the builtins though)

 * A second approach would be to make the `flux(1)` cmd driver itself emit the list of possible subcommands. A hidden  builtin command or an option could be added which would print all possible subcommands given the current environment and builtin paths. Eventually `optparse` itself and Python argparse could be extended with similar options, so that commands at each level could list subcommands (and options) in a standard way (I think something like this was already suggested by @SteVwonder)

garrettbslone replied:
Thanks for the suggestions on generating the subcommands dynamically! 

However, when I generate the list of subcommands and builtins, I get some discrepancies between the generated commands and the commands listed by flux help. 

The list of subcommands (first row) and builtins (second row) generated:
<img width="824" alt="image" src="https://user-images.githubusercontent.com/37486957/77836216-7cc6af80-7111-11ea-816e-a111edfe3ced.png">

The list of commands from flux help:
<img width="630" alt="image" src="https://user-images.githubusercontent.com/37486957/77836324-8c92c380-7112-11ea-8dbb-a6c2ab284ca0.png">

For example, from the subcommands, aggregate, jobspec, job, queue, and comms are all not listed under flux help. Does anyone have any suggestions/things I may be missing?

SteVwonder replied:
Very cool that you got the auto-generation working!

I think it is intended that some commands be "blacklisted" from the `flux --help` listing.  I'm surprised though that `job` and `queue` aren't already in the `--help` listing. Hmmmm.

Maybe it is a separate issue as to which commands exactly are blacklisted, but I think as long as the blacklist functionality exists, we can tune and tweak it later.

grondo replied:
`flux help` reads the list of commands from `$(datadir)/flux/help.d/*.json`. For flux-core, the JSON input is autogenerated by `etc/gen-cmdhelp.pl`, which looks through the `*.adoc` manpages looking for a `flux-help-include: true` tag. This way, as commands are added/removed they automatically appear/disappear from `flux help`. The drawback is if there is no manpage, then the command won't appear in `flux-help`.

We'll probably want to expand this system to have different "levels" of help in the near future (perhaps displaying different results for owner vs guest if possible.)

SteVwonder commented:
Yep! The extglob shopt fixed it for me.  I think you can probably add that directly to the script.  It seems to be what bash-completion does in ubuntu: [[1]](https://askubuntu.com/questions/634397/where-is-bashs-shopt-extglob-turned-on-for-my-interactive-shell) [[2]](https://github.com/scop/bash-completion/blob/master/bash_completion#L45)

Unfortunately, I don't know what the side-effects of enabling it are or how this might break users.

garlick commented:
Following up on our conversation today:

I was going to find an automake subtitution variable already defined in Flux's build system for you to use.  Well this works: put this in your `foo.in`
```
exec_prefix=@X_EXEC_PREFIX@
fluxcmddir=@fluxcmddir@
```
and it gets expanded in `foo` to :
```
exec_prefix=/usr/local
fluxcmddir=${exec_prefix}/libexec/flux/cmd
```
You would add your script to AC_CONFIG_FILES in `configure.ac` (without the .in).
 
But actually, maybe it would be eaiser to skip the automake subsitution and just run something like this in your script?
```console
$ ls -1 $(flux env printenv FLUX_EXEC_PATH)
```

garlick commented:
@garrettbslone - when you're ready for a review, would you mind summarizing what's changed since the last round of comments?  Thanks!

garrettbslone commented:
Since the last round of comments, I added a script get_builtins.sh in flux-core/etc/completions that generates a list of the builtin commands at build time and copies them to the command completion script. I also added the command completion script flux-core/etc/completions/flux to be installed in the makefile. 

However, the build is failing in Travis when trying to run:
`make --output-sync=target --no-print-directory -k -j 2 distcheck`

After seeing this I tried running: `make distcheck -d` on my machine.

Which was also failing. Both times the build is failing when it enters flux-core/etc with the error:
`make[2]: *** No rule to make target 'completions/flux', needed by 'all-am'.  Stop.`

I looked inside of the newly generated build tree and didn't see any of the files I committed. If anyone has any ideas on how to fix this that'd be great.

garlick commented:
The distcheck target does a [vpath build](https://www.gnu.org/software/automake/manual/html_node/VPATH-Builds.html), where the build directory is separate from the source directory.  You can do one manually like this:
```
$ cd flux-core
$ mkdir _build
$ cd _build
$ ../configure
$ make -j4
...
Making all in etc
make[1]: Entering directory '/home/garlick/proj/flux-core/_build/etc'
  GEN      flux/help.d/core.json
  GEN      flux/curve
Saving /home/garlick/proj/flux-core/etc/flux/curve/client
Saving /home/garlick/proj/flux-core/etc/flux/curve/client_private
Saving /home/garlick/proj/flux-core/etc/flux/curve/server
Saving /home/garlick/proj/flux-core/etc/flux/curve/server_private
chmod +x ./completions/get_builtins.sh && \
./completions/get_builtins.sh
chmod: cannot access './completions/get_builtins.sh': No such file or directory
Makefile:886: recipe for target 'all' failed
make[1]: *** [all] Error 1
make[1]: Leaving directory '/home/garlick/proj/flux-core/_build/etc'
Makefile:580: recipe for target 'all-recursive' failed
make: *** [all-recursive] Error 1
```
To run a script from the Makefile so that it works in this environment, you'll need to put `$(srcdir)` in front of it so that it is accessed in the source directory not the current directory (which is the build directory).

The script has the path to its output hardwired, and that will have a similar problem, so I would just have the script produce its output on standard output, and redirect it from the Makefile where you can specify the path more easily.

One other problem is the `get_builtins.sh` script modifies the `competions/flux` script which is checked into git, so that after the script runs, the git working directory will have unstaged changes.   
I think you'll want to combine `completions/flux` with the output of `get_builtins.sh` to generate a new script that is installed, like what's [described here](https://www.gnu.org/software/automake/manual/html_node/Scripts.html).

garlick commented:
Ooh, tests are passing.  Ready for another review pass?

garrettbslone commented:
I think so. Since the last round of commits, I moved `completions/flux` to `completions/flux.in` and have the `completions/flux` file generated from `configure.ac`. I also changed `get_builtins.sh` to take an arg for the bath to the src directory instead of hardcoded. Lastly, I moved `get_builtins.sh` from noinst_SCRIPTS to EXTRA_DIST in the makefile.

garlick commented:
OK, checking it out.  In the future you may want to create a topic branch instead of using your master for development.  This is OK for this one though!

garlick commented:
Nice!  I installed to `/usr/local` on my test system and sourced `/usr/local/etc/bash_completion.d/flux`, then verified
* `flux lsattr` (a builtin)
* `flux kvs` (an installed command)
* `flux plumbus` (an executable I dropped in `/usr/local/libexec/flux/cmd` after the fact, as if it were a command supplied by another project like flux-sched)
* then I ran `/usr/local/bin/flux start` and repeated the above - all good!
* then I ran `./flux start` from source tree and repeated the above - as expected `flux plumbus` not found until I dropped it into `./src/cmd`

I did run into a couple of minor things
* my installed `flux` script seems to have two `FLUX_BUILTIN=` lines at the end
* some trailing whitespace flagged in red when you use `git diff` to view your changes
* the chmod +x  on `get_builtins.sh` (a checked in file) causes unstaged changes.  You'll want to change the mode in git with e.g. `git update-index --chmod=+x file` instead

I think you can go ahead and squash down the commits into something like
* add `get_builtins.sh` script
* generate bash completion script

Each commit should have a paragraph or two explaining what the change does, and the completion would could say `Fixes #1647` to auto-close that bug.  Have a look at `git log` to see the general form for commits.

Nice work!

SteVwonder commented:
Thanks @garrettbslone for putting this together.  I'm still getting the syntax error from before:
```
[herbein1@quartz1148:~/opt/packages/toss3/flux-core/bash-completion]
[20:02:38] $ source etc/bash_completion.d/flux
bash: etc/bash_completion.d/flux: line 26: syntax error near unexpected token `('
bash: etc/bash_completion.d/flux: line 26: `            @(-*))'
```

I think you'll need to put the `shopt` up above the `_flux_core` function.  That fixes it for me.

Other than that LGTM! Thanks again!

garlick commented:
> I think you'll need to put the shopt up above the _flux_core function.

I take it this is an issue with the older bash version TOSS?  (Just wondering why I missed that)

@garrettbslone - you should add yourself to the top level AUTHORS file in a separate commit!

grondo commented on etc/Makefile.am:
```diff
@@ -40,5 +40,20 @@ pkgconfig_DATA = flux-core.pc \
+	completions/get_builtins.sh \
 	gen-cmdhelp.pl
+
+all: 
+	chmod +x $(srcdir)/completions/get_builtins.sh && \
+	$(srcdir)/completions/get_builtins.sh  $(top_srcdir)/src/cmd/builtin/ \
+	>> completions/flux
```
In the automake rule above

 * You shouldn't need to add execute permission to the `get_builtins.sh` script every time. Just make sure the script is checked into git with execute permission...

 * The `all:` rule is run every time, and appends to the `completion/flux` file, so if you run `make` more than once the file gets multiple FLUX_BUILTINS appended to the end.

 * It would be tricky to create a Makefile rule for `completions/flux` since it is already generated by `configure`. I wonder if it would be better to generate `completions/flux.pre` or similar from `configure`, then combine it and output of `get_builtins.sh` to create `completions/flux`, so a rule for `completions_flux` can be used instead of `all`.

Hope that makes sense!

garrettbslone commented:
Changes since last time: renamed the base completions file from `flux.in` to `flux.pre.in`. Added the completion script to `CLEANFILES`, and changed the rule to make `completions/flux` to its own rule instead of all.

garlick commented:
Thanks @garrettbslone.  I think we're getting close!  Couple quick comments:

I think my advice to split `get_builtins.sh` out to its own commit was probably dumb.  Feel free to squash it into the other one.

Generating `flux.pre` from `flux.pre.in` seems to be a no op?  (Just a copy?)  Might as well just check in `flux.pre` instead.

grondo commented:
> Generating flux.pre from flux.pre.in seems to be a no op? (Just a copy?) Might as well just check in flux.pre instead.

Sorry, that might have been my fault (see third bullet in [this comment](https://github.com/flux-framework/flux-core/pull/2755#discussion_r429332779) above)

We want to create a Makefile rule that will regenerate the final completions file when any of the dependencies change. I think I may have confused @garrettbslone with my suggestion.

grondo commented:
btw, for me this branch fails when built in a vpath build, e.g.
```
~/flux-core.git $ make distclean
~/flux-core.git $ mkdir b
~/flux-core.git $ cd b
~/flux-core.git/b $ ../configure
[snip]
~/flux-core.git/b $ make
[snip]
Making all in etc
make[1]: Entering directory '/home/grondo/git/f.git/b/etc'
make[1]: *** No rule to make target '../../etc/completions/flux.pre', needed by 'completions/flux'.  Stop.
make[1]: Leaving directory '/home/grondo/git/f.git/b/etc'
Makefile:583: recipe for target 'all-recursive' failed
make: *** [all-recursive] Error 1



```

grondo commented:
Finally, as an added bonus, prefix your new rule commands with `$(AM_V_GEN)` (e.g. see the rule for `flux/curve`). That way in default "quiet" build mode make emits:
```
  GEN      completions/flux
```

instead of 
```
test -d completions || mkdir completions && \
cp completions/flux.pre completions/flux && chmod +w completions/flux && \
../../etc/completions/get_builtins.sh \
../../src/cmd/builtin >> completions/flux

```

garrettbslone commented:
The `flux.pre.in` to `flux.pre` was a misunderstnding of @grondo's earlier comment. 

> Did you run into problems with `cp` creating read-only files that required the `chmod +w`? I would not think that would normally be necessary.

Wrt the `cp`, it was creating `completions/flux` as read only. I did try adding the `-p` and `-a` flags (together and seperate), with no luck.

grondo commented:
> Wrt the cp, it was creating completions/flux as read only. I did try adding the -p and -a flags (together and seperate), with no luck.

Hm, not sure what was going on there.

Here's another thought. The `cp` could be avoided by generating the "builtins" stanza in its own rule, then combining the checked in `completions/flux.pre` and the generated `completions/flux.builtin` to create `completions/flux`:
```diff
 diff --git a/etc/Makefile.am b/etc/Makefile.am
index de287d54f..5c3146de3 100644
--- a/etc/Makefile.am
+++ b/etc/Makefile.am
@@ -32,7 +32,7 @@ endif
 clean-local:
        -rm -rf flux
         
-CLEANFILES=completions/flux
+CLEANFILES=completions/flux completions/flux.builtins
 
 if WITH_PKG_CONFIG
 pkgconfig_DATA = flux-core.pc \
@@ -51,11 +51,12 @@ EXTRA_DIST = \
        gen-cmdhelp.pl \
        $(noinst_SCRIPTS)
 
-completions/flux: $(srcdir)/completions/flux.pre
-       test -d completions || mkdir completions && \
-       cp $< $@ && chmod +w $@ && \
-       $(srcdir)/completions/get_builtins.sh \
-       $(top_srcdir)/src/cmd/builtin >> $@
+completions/flux.builtins: $(srcdir)/completions/get_builtins.sh
+       $(AM_V_GEN)$(srcdir)/completions/get_builtins.sh \
+           $(top_srcdir)/src/cmd/builtin > $@
+
+completions/flux: completions/flux.pre completions/flux.builtins
+       $(AM_V_GEN)cat $^ > $@
 
 fluxcompdir = $(sysconfdir)/bash_completion.d
 fluxcomp_SCRIPTS = \

```


Not sure if that is a net win or not

garrettbslone commented:
Changes: Removed `flux.pre.in` and replced with `flux.pre`.

grondo commented on etc/Makefile.am:
```diff
@@ -40,5 +42,20 @@ pkgconfig_DATA = flux-core.pc \
+	$(noinst_SCRIPTS)
+
+completions/flux: $(srcdir)/completions/flux.pre
+	$(AM_V_GEN)test -d completions || mkdir completions && \
```
I still don't think you need to `mkdir completions` when `completions/flux.pre` already exists (and therefore its parent directory) since that file is an explicit prerequisite of this rule.

garrettbslone replied:
The `mkdir completions` is necessary for a vpath build since the `completions/flux.pre` isn't being installed, e.g. removing that line from the `etc/Makefile` fails a vpath build:
```bash
Making all in etc
make[1]: Entering directory '/home/garrett/flux-core/b/etc'
  GEN      completions/flux
  GEN      flux/curve
  GEN      flux/help.d/core.json
cp: cannot create regular file 'completions/flux': No such file or directory
make[1]: *** [Makefile:893: completions/flux] Error 1
```

grondo replied:
Yeah, sorry I missed that `completions/flux.pre` is in srcdir.

garrettbslone commented:
From the last round of changes (which did include rebase-ing to include hwloc 2.0 changes) checks are failing on the hwloc tests. Anyone have any thoughts on what might be causing this?
```make
FAIL: t2005-hwloc-basic.t 19 - hwloc: reload with no args reloads system topology
```

grondo commented:
Can you run `./t2500-hwloc-basic.t -d -v` to get extra debug?

garrettbslone commented:
Here is the part that's failing:
```
expecting success: 
    flux hwloc reload &&
    flux hwloc topology > system.out4 &&
    test_cmp system.topology.out system.out4

Invalid MIT-MAGIC-COOKIE-1 keyInvalid MIT-MAGIC-COOKIE-1 key--- system.topology.out	2020-06-05 23:01:30.489292658 +0000
+++ system.out4	2020-06-05 23:01:30.925290892 +0000
@@ -151,10 +151,10 @@
           </object>
         </object>
       </object>
-      <object type="Bridge" bridge_type="1-1" depth="1" bridge_pci="0000:[08-08]" pci_busid="0000:00:03.1" pci_type="0604 [1022:1483] [1043:87c0] 00" pci_link_speed="4.000000">
+      <object type="Bridge" bridge_type="1-1" depth="1" bridge_pci="0000:[08-08]" pci_busid="0000:00:03.1" pci_type="0604 [1022:1483] [1043:87c0] 00" pci_link_speed="8.000000">
         <info name="PCIVendor" value="Advanced Micro Devices, Inc. [AMD]"/>
         <info name="PCIDevice" value="Starship/Matisse GPP Bridge"/>
-        <object type="PCIDev" pci_busid="0000:08:00.0" pci_type="0300 [10de:1e84] [1462:3860] a1" pci_link_speed="4.000000">
+        <object type="PCIDev" pci_busid="0000:08:00.0" pci_type="0300 [10de:1e84] [1462:3860] a1" pci_link_speed="8.000000">
           <info name="PCIVendor" value="NVIDIA Corporation"/>
           <info name="PCIDevice" value="TU104 [GeForce RTX 2070 SUPER]"/>
           <object type="OSDev" name=":1.0" osdev_type="1">
@@ -358,10 +358,10 @@
           </object>
         </object>
       </object>
-      <object type="Bridge" bridge_type="1-1" depth="1" bridge_pci="0000:[08-08]" pci_busid="0000:00:03.1" pci_type="0604 [1022:1483] [1043:87c0] 00" pci_link_speed="4.000000">
+      <object type="Bridge" bridge_type="1-1" depth="1" bridge_pci="0000:[08-08]" pci_busid="0000:00:03.1" pci_type="0604 [1022:1483] [1043:87c0] 00" pci_link_speed="8.000000">
         <info name="PCIVendor" value="Advanced Micro Devices, Inc. [AMD]"/>
         <info name="PCIDevice" value="Starship/Matisse GPP Bridge"/>
-        <object type="PCIDev" pci_busid="0000:08:00.0" pci_type="0300 [10de:1e84] [1462:3860] a1" pci_link_speed="4.000000">
+        <object type="PCIDev" pci_busid="0000:08:00.0" pci_type="0300 [10de:1e84] [1462:3860] a1" pci_link_speed="8.000000">
           <info name="PCIVendor" value="NVIDIA Corporation"/>
           <info name="PCIDevice" value="TU104 [GeForce RTX 2070 SUPER]"/>
           <object type="OSDev" name=":1.0" osdev_type="1">
not ok 19 - hwloc: reload with no args reloads system topology
#	
#	    flux hwloc reload &&
#	    flux hwloc topology > system.out4 &&
#	    test_cmp system.topology.out system.out4
#	
```

grondo commented:
How bizzarre. The link speed is changing between runs?

garrettbslone commented:
> How bizzarre. The link speed is changing between runs?

It looks like it, but it only seems to be failing sometimes. I rebuilt and worked fine, but rebuilt again and it failed again.

grondo commented:
I haven't seen this problem before, but it seems that, on your system at least, the underlying test is not stable. I wonder if there is a way to filter out these device attributes that might change between subsequent runs.

garlick commented:
It sounds like Nvidia may have a power saving feature that adjusts PCI bus speed in response to demand?

http://billauer.co.il/blog/2018/08/nvidia-low-pci-express-speed/

grondo commented:
For these systems, I wonder if we need a version of `flux hwloc reload` that filters out these kinds of devices?

garlick commented:
@garrettbslone would you mind summarizing what you changed today?  Was it just a rebase?

garrettbslone commented:
Just a rebase over master.

garlick commented:
Just tested and the finished product works when installed to `/etc/bash_completion.d`.

There are some trailing spaces/tabs in scripts (have a look at your commit with git show and it should be highlighted in red).

grondo commented on etc/Makefile.am:
```diff
@@ -40,5 +42,20 @@ pkgconfig_DATA = flux-core.pc \
+	$(srcdir)/completions/get_builtins.sh \
+	$(top_srcdir)/src/cmd/builtin >> $@
+
+fluxcompdir = $(sysconfdir)/bash_completion.d
```
I noticed on my Ubuntu system that there is a `bash-completion.pc` pkg-config file, and that it shows:
```
$ cat /usr/share/pkgconfig/bash-completion.pc 
prefix=/usr
compatdir=/etc/bash_completion.d
completionsdir=${prefix}/share/bash-completion/completions
helpersdir=${prefix}/share/bash-completion/helpers

Name: bash-completion
Description: programmable completion for the bash shell
URL: https://github.com/scop/bash-completion
Version: 2.8
```

Seems like `fluxcompdir` should be set via pkg-config lookup for `bash-completion`, falling back to `$(sysconfdir)/bash_completion.d` if not found.

garlick replied:
Oh good find!

> Seems like `fluxcompdir` should be set via pkg-config lookup for `bash-completion`, falling back to `$(sysconfdir)/bash_completion.d` if not found.

Hmm, but that wouldn't be right for a side-install, e.g. `make install` would fail with permission denied for, say, `--prefix=$HOME/local`.

Maybe we should go ahead and merge this, but open an issue to discuss the right way to use `bash-completion.pc`.  Off the top of my head I'm not having any great ideas...

grondo replied:
> Hmm, but that wouldn't be right for a side-install, e.g. make install would fail with permission denied for, say, --prefix=$HOME/local.

Ah, I hadn't considered that. My main discovery was that `$(sysconfdir)/bash_completion.d` is the "compat" directory, and the real bash completion path is `${prefix}/share/bash-completion/completions`

grondo replied:
Hm, apparently you can do this with `pkg-config`:
```
$ pkg-config --define-variable=prefix=/usr/local --variable=completionsdir bash-completion
/usr/local/share/bash-completion/completions
```

grondo commented:
Sounds like consensus is to merge, so I'll set MWP.

I'll open an issue on the question about use of pkg-config to get the `fluxcompdir`

grondo commented:
Ooh, thanks for remembering to fix the PR title @garlick!

SteVwonder commented:
It looks like mergify is unhappy (maybe because the config changed between when this PR opened and now).  It also looks like the flux-sched check hung and timedout.  Going to try and rebase with mergify and see if that clears everything up.

SteVwonder commented:
@mergifyio rebase

SteVwonder commented:
Hmmm.  Maybe it's just upset about the Github CI timing out?  Either way, I kicked the Github CI.  We can poke Mergify with a refresh once that comes back clean.
